### PR TITLE
361 taskmeasureinstrument handling with document upload and status change

### LINKED ISSUE #361: Task/Measure/Tool handling with document upload and status change
## Description
We have build a modal in which we can edit measures in a textbox and change the status with a dropdown. 
In this ticket we will add some features to the modal and extend the situations in which it is used. 

**Situations to work on a task**:
1. **Updating a measure**: This is the same situation as before.
3. **Updating an hulpmiddel/tool**:  _New situation_ in which we handle an entire instrument as one task, instead of a task per instrument question. (for now out of scope that the actual hulpmiddel/tool is imported, also the buttons from create new algorithm out of scope).

**Features to work on a task**:   

1.. Upload document possibilities. No maximum number of files for now. Support for every possible extension.
- In addition to uploading, you could also include a reference to a file located in a samenwerkruime, for example. This would essentially be a URI. Design to be made. 
- If upload is not successful there needs to be a feedback to the user. Instead of the text being 'Dit is een foutmelding' the text needs to be 'Fout bij uploaden document'/'Document upload error'. 
- With minio encrypt files per default
- icons of the filetype extension for already uploaded files (use special in CSS selector ($=*.jpeg)).
- You can remove the file from system (both before the uploaded file is uploaded or when it is uploaded).
- Only upload when you press upload button.
- Metadata (link to algorithm, timestamp, user who exported).
- Authorization, who can see it, can download the files. 
- Progress bar file upload (example: https://htmx.org/examples/file-upload/)

2. Add 'In review'/'Beoordelen' and 'Skip'/'Overslaan' to the possible statuses. 

## Design of the modal with all features: 
Image
Image
Image

Na opslaan en weer openen
Image

Foutmelding
Image

### Out of scope:
- Owner, Action holder are roles on a measure which can be assigned. Option to choose from all team members that are part of this algorithm. 
-  Add the Roles that should do the measure from AK. Roles can be handled in the same way as the titles and description. 
- There are many measures that have the same or very similar _title_ as the _description_. For now we do not want to include the _explanation_ by default as it can contain a lot of text and therefore makes it messy. We stick with the linking to AK for more info. 
- Thread with mentions and comments 
- Description field from CKEditor 
- Add examples of how the measure was executed where possible (needs to come from AK)
- Timetravelling (go back to previous version) 
- Add a description field for a discussion in the most basic way: user, comment, date, and option to edit or delete it. No reply, just 1 stream of comments. 
- Include a log of all edit dates and editors 
- Refactoring of the system card of `measures` and `assessment cards`
- Add last-edited timestamp and by who it was edited. The last edit counts for all edits done on this page. 

## Technical Implementation
- [ ] Error handling in POST endpoint to make sure we cleanup on `/algorithm/<ID>/file`
- [ ] We write temporary files to local file storage
- [ ] We call the s3 minio endpoint to store the file in persisted storage
- [ ] MinBZK/ai-validation-infra#67
- [ ] We store our files here:
  - path: `uploads/org/<ORG-ID>/algorithm/<ALG-ID>/<ULID>`
  - metadata for the file:
    - ALG-ID
    - uploader user-id
    - timestamp
    - filename
    - extension
- [ ] We store the whole path in the system-card after persisting to minio after each file that succeeded
- [ ] In the frontend we show the filename and an icon based on the metadata from minio
- [ ] We create a single encryption key that we store as SOPS secrets so that we have it available as an env var
- [ ] We create a GET/DELETE endpoint for ULIDs `/algorithm/<ID>/file/<ULID>`

### PR DESCRIPTION
**Pull request by ChristopherSpelt**
# Description

## What is added
1. Added additional statuses `(todo, in progress, in review, done, not implemented)`. Note that this also means adding additional colours to the status bar icon in the measures in the requirements overview. Since the status bar only has a limited set of colours, this is now suboptimal visually. The current colours are aligned with Ruben.
2. Possibility to add URI's to files. The input field is not checked, so any user input will be stored in the system card under the corresponding measure in the field `links`.
3. Possibility to upload files. NOTE: it was decided to REMOVE the file list when adding files, before saved is pressed.
4. Possibility to download an uploaded file from the object storage.
5. Possibility to remove the uploaded files. 
6. Added tests for the new ObjectStorageService
7. Added tests for the new GET and DELETE endpoints for the files

## What is not added 
1. Automatically render an icon according to the file extension. For example a PDF logo for a PDF file.
2. A progress bar for uploading files.

## What still needs to be done
1. (very OPTIONAL) remove the coupling of the ObjectStorageService with Minio, maybe by using protocols or something.
2. Code review
3. Code review + merge of the PR in the infra repo: https://github.com/MinBZK/ai-validation-infra/pull/85
4. Check if everything works on staging
5. Merge the PR

Estimation of leftover points: 1 or 2.

## If things don't work properly 
-> Berry probably knows most about the Minio side of things
-> Robbert U knows probably most about the way the UI of the form is implemented

## How to tests things locally
Use a local instance of Minio with the settings which are specified in the core settings in AMT.

Resolves #361 

## Checklist

Please check all the boxes that apply to this pull request using "x":

-   [x] I have tested the changes locally and verified that they work as expected.
-   [x] I have followed the project's coding conventions and style guidelines.
-   [x] I have rebased my branch onto the latest commit of the main branch.
-   [x] I have squashed or reorganized my commits into logical units.
-   [x] I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

### CODE REVIEW

File: amt/locale/nl_NL/LC_MESSAGES/messages.po [hunk 1]
```diff
@@ -310,6 +342,12 @@ msgstr ""
 "Er is iets fout gegaan tijdens de autorisatiestroom. Probeer het later "
 "opnieuw"
 
+#: amt/core/exceptions.py:84
+msgid "Something went wrong storing your file. PLease try again later."
+msgstr ""
+"Er is iets fout gegaan tijdens het opslaan van uw bestand. Probeer het "
+"later opnieuw"
```
Comment (laurensWe) [inline amt/locale/nl_NL/LC_MESSAGES/messages.po:349]:
```suggestion
"later opnieuw."
```

File: amt/locale/nl_NL/LC_MESSAGES/messages.po [hunk 2]
```diff
@@ -654,22 +663,34 @@ msgstr "Er is één fout:"
...
+msgid "Saved files"
+msgstr "Opgeslagen bestanden"
+
+#: amt/site/templates/macros/form_macros.html.j2:150
+msgid "Delete file"
+msgstr "Verwijder bestand"
+
+#: amt/site/templates/macros/form_macros.html.j2:157
+msgid "Are you sure you want to delete"
+msgstr "Weet u zeker dat u het bestand wilt verwijderen "
```
Comment (laurensWe) [inline amt/locale/nl_NL/LC_MESSAGES/messages.po:692]:
```suggestion
msgstr "Weet u zeker dat u het bestand wilt verwijderen"
```

**Review by laurensWe — DISMISSED** [review]

**Review by ravimeijerrig — APPROVED** [review]
LGTM :)

**Review by ravimeijerrig — APPROVED** [review]
LGTM :)

**Pull request merged**